### ISSUE TITLE
Errors when watcher backend is given broken symlinks

### ISSUE THREAD

**Issue by sunng87** [passcod/rsnotify#21]
I'm on Linux kernel 4.1.0, rsnotify 2.1.0 and Emacs 24.5.

My code is as simple as:

```rust
fn _watch (p: &Path) -> Result<(), Error>{
    let (tx, rx) = channel();
    let mut watcher: RecommendedWatcher = try!(Watcher::new(tx));
    println!("watcher init");
    try!(watcher.watch(p));
    println!("watching");
    let _ = rx.recv();
    Ok(())
}
```

Every time I modified files in Emacs, `watcher.watch(p)` returns error:

```
Io(Error { repr: Os { code: 2, message: "No such file or directory" } })
```

If I `ls -al` the watched directory, I'm getting:

```
 $ ls -al examples/templates  
drwxr-xr-x 2 nsun nsun 4096 Jul  6 21:04 .
drwxr-xr-x 3 nsun nsun 4096 Jul  4 10:37 ..
lrwxrwxrwx 1 nsun nsun   31 Jul  6 21:04 .#index.hbs -> nsun@thinkless.17462:1435199339
-rw-r--r-- 1 nsun nsun  279 Jul  6 20:57 index.hbs
```

The `.#index.hbs` symbol link, as [the SO question](http://stackoverflow.com/questions/5738170/why-does-emacs-create-temporary-symbolic-links-for-modified-files) said, is Emacs lockfile. 

I'm not quite sure if this is an rsnotify issue, anything suggestion or workaround are welcomed.

**Comment by JIghtuse** [comment]
It also happens if `watcher.watch` encounters broken symlink. Try to create one and you will see the same result. rsnotify just takes what walker gives it and throws it into `add_watch()`. What should it do instead?

**Comment by passcod** [comment]
It should probably check that the path it's given is a directory or a regular file, and if it's a symlink it should attempt to dereference it, and only feed it to the watcher if either of those is the case. I imagine rsnotify also fails if invoked on, say, "special" files in `/dev`, `/sys`, or `/proc`.

**Comment by passcod** [comment]
Fix shipped in v2.5.2.
